### DISCUSSION TITLE
how can i transform the ast to code but not compile it

### DISCUSSION THREAD

**Discussion by lizuncong** [handlebars-lang/handlebars.js#1826] [Q&A]
Before filing issues, please check the following points first:

- [ ] Please don't open issues for security issues. Instead, file a report at https://www.npmjs.com/advisories/report?package=handlebars
- [ ] Have a look at https://github.com/handlebars-lang/handlebars.js/blob/master/CONTRIBUTING.md
- [ ] Read the FAQ at https://github.com/handlebars-lang/handlebars.js/blob/master/FAQ.md
- [ ] Use the jsfiddle-template at https://jsfiddle.net/4nbwjaqz/4/ to reproduce problems or bugs

This will probably help you to get a solution faster. 
For bugs, it would be great to have a PR with a failing test-case.

**Comment by jaylinski** [comment]
What exactly are you trying to do?

**Reply by lizuncong** [reply]
```html
{{#my_import 'test-section'}}{{/cli_import}}
<section>
    {{#if isActive}}
       <div>test</div>
    {{/if}}
    {{ color }}
</section>
```
Suppose the test-section is as follows
```html
<div>my test section</div>
```

my_import is a custom helper for inserting reusable code snippets.
In my business scenario, when I call `handlebars.compile`, I hope to only compile my_import, and other helpers, such as if, etc. are output as strings.
I use ast to modify nodes, convert other helpers to strings, except my_import. 
I expect the end result to be as follows

```html
<div>my test section</div>
<section>
    {{#if isActive}}
       <div>test</div>
    {{/if}}
    {{ color }}
</section>
```


```js
const handlebars = require('handlebars')

module.exports = function transformAst(program) {
    const Visitor = handlebars.Visitor

    class Scanner extends Visitor {
        mutating = true

        BlockStatement(node) {
            const {path: _path, params, loc, program: { body }} = node
            const helperName = _path.original

            if (helperName.indexOf('my_') > -1) {
                return node
            }
            const newBody = body.reduce((str, cur) => {return cur.original + str}, '')
           // The question is, how can I ensure the location of the source code???
            const newValue =  `{{#${helperName} ${params[0].original}}}${newBody}{{/${helperName}}`

            const newNode = {
                "type": "ContentStatement",
                "original": newValue,
                "value": newValue,
                "loc": loc
            }
            return newNode
        }
    }

    const scanner = new Scanner()

    scanner.accept(program)
}
```

In this way, I can indeed achieve my needs, but the location information of the source code is lost.

**Reply by lizuncong** [reply]
@jaylinski hello, is there any way to solve this problem?

**Reply by jaylinski** [reply]
I'm afraid there is no easy way to do this.

**Comment by nknapp** [comment]
So, you want to recreate the original template from the AST?

**Reply by lizuncong** [reply]
Yes, I just want to compile a part of the helper, don't compile the rest of the helpers, the rest of the helpers are output as they are
